### ISSUE TITLE
Documentation: Automatic Differentiation to Compute the Adjoint Field?

### ISSUE THREAD

**Issue by QuentinWach** [ymahlau/fdtdx#58]
Am I understanding correctly that you are not running two FDTD simulations but only one and then you "backpropagate" through the computational graph of the simulation to get the adjoint?

As I understand it, this approach is well known to not scale well due to memory issues and your contribution is to lessen this problem in various clever ways. Would you mind to go into **more details in the documentation** or point me to **related articles** that explain this? 

Also, it seems like, if this is indeed true, that this full autodiff approach is still a leading issue that causes FDTDx to lag behind FlexCompute's Tidy3D for very large simulations. After all, they seem to only be using autodiff to update the topology and geometry rather than finding the adjoint field where they simply run two manual FDTD simulations (one forward, one backward). Would you agree?

**Comment by ymahlau** [comment]
Hi, that's a great question and we definitely need to document this a bit more :)

Regarding backpropagation, you are correct for the simple way of implementing backpropagation. If you run the [forward function](https://github.com/ymahlau/fdtdx/blob/4344fdc81a0121b535e1122f6e9efc8e75d121c2/src/fdtdx/fdtd/forward.py#L124), which computes a single FDTD time step, in a loop, and then wrap you function in jax.grad, you would indeed only perform the forward simulation and backpropagate through the computational graph. 

The problem, as you mentioned, with this approach is that it is necessary to store intermediate values to perform the backpropagation. Specifically, it is necessary to store the E and H field of the entire simulation volume after every time step. This of course quickly explodes the capabillities of any reasonable hardware setup.

One way to remedy this, is to use [checkpointing](https://github.com/ymahlau/fdtdx/blob/4344fdc81a0121b535e1122f6e9efc8e75d121c2/src/fdtdx/fdtd/fdtd.py#L378). This only saves the E and H field only at a constant number k time steps and recomputes the missing values in during backpropagation by performing the corresponding FDTD simulations from the last saved checkpoint again. Additionally, there are algorithms that compute where the optimal time steps for saving these checkpoints are and how the memory can be reused if a checkpoint is no longer needed. But, this is still very slow for a large number of time steps or low number of checkpoints.

Since Maxwell's equations and in extension FDTD are reversible in time, it is also possible to compute an "inverse" FDTD function, which takes the E and H field at time t and computes the E and H field at time t-1. This alleviates the memory issues, because the E and H field do not need to be saved anymore. Instead, one computes the forward pass, and starting from the last time step, computes the previous time steps and the performs the backpropagtion through the computational graph. This is the what the [reversible_fdtd](https://github.com/ymahlau/fdtdx/blob/4344fdc81a0121b535e1122f6e9efc8e75d121c2/src/fdtdx/fdtd/fdtd.py#L16) function uses. So basically this performs a forward FDTD function and then an interleaved time reversed FDTD function and backpropagation ( I think from a runtime perspective this is probably like 3-4 FDTD simulations because backpropagation also needs some time). 
The only caviat here is that the PML is not invertible, because energy is absorbed and therefore information lost. As a consequence, we save the E and H field at every time step, but only for the 2D interface between PML and simulation volume, which needs much less memory than the full 3D simulation volume. 

For the adjoint method, I have to admit that I am not an expert because I come from a CS background, so if anything what I say about it is wrong please correct me. Using the adjoint method, it is necessary to save the E and H field during the forward FDTD function at every time step as well, but only for the 3D volume of the device geometry. This means that depending on the size of the geometry that needs to be optimized, this may lead to a larger or smaller memory footprint compared to the 2D interface slices saved using our current method. I would think that the 2D slices need less memory for most applications. But, you are correct that the adjoint method is faster, because it does not need the backpropagation step. 

In the applications that I have usually memory is the bigger issue compared to runtime, because in the end you can just let the simulation run for a bit longer, but if you need more memory you actually have to buy new hardware. For Tidy3D this of course is not an issue, because they probably have a large cluster of GPUs that they can just allocate to jobs as they see fit.

I think for the time comparison you are probably reffering to table 2 in our [whitepaper](https://arxiv.org/pdf/2412.12360). In this experiment, we only compared the speed of the forward simulation, so the adjoint method does not contribute to the speed difference. I believe the main reason for the speed difference between FDTDX and Tidy3D in this experiment is that Jax does not work so well with a multi-gpu setup. This leads to a lot of communication overhead. Tidy3D used custom CUDA-Kernels and a smart way to minimize communication, but this is only possible to do if you have a large team of engineers. We do think that we can make some substantial speed improvements to FDTDX, but Tidy3D is probably always going to be a bit faster, simply because they have large team of engineers instead of a single PhD-Student :)

I will probably put this long essay that I just wrote somewhere in the documentation.

As a last point, I am currently working on a paper which presents an exhaustive comparison and overview of the methods that I just described. 

related to #55

**Comment by QuentinWach** [comment]
I have been thinking about this a lot and still am, since, admittedly, I am also working on my own implementation for FDTD sims and inverse design for work and my M.Sc. thesis, though I am running the adjoint method manually without auto-diff but simply by calculating the overlap. So **thank you** for the detailed response! I will have to look more deeply into the time-reversal approach. It seems both very clever but also very alien still. I can see how you'll quickly run into issues when trying to incorporate other effects like absorption or nonlinear responses.

I should also correct myself: Like you said, Tidy3D is using custom CUDA-Kernels in the backend. Their auto-diff options are apparently a separate option, built using Jax as well (but without your innovations as far as I can tell), and purely optional and experimental.

I know it isn't perfectly aligned with your incentives as a PhD student to give so much to the community without a paper as a reward, so thank you for the additions you're making. And I am really looking forward to the paper your planning :D.
